README.md
# Ocular Immune-Mediated Inflammatory Diseases Ontology (**OcIMIDo**)

#### See :grey_exclamation: [Issues](https://github.com/sap218/ocular-immune-mediated-inflammatory-disease-ontology/issues)  for the term-tracker to submit a request (new term/change/suggestion) 


Following the MIRO guidelines for reporting on an ontology [1].

---

**A. The basics**
  * A.1 Ocular Immune-Mediated Inflammatory Diseases Ontology (OcIMIDo) v1.2.0.
  * A.2 Ontology owner: [Samantha C Pendleton](mailto:samanfapc@gmail.com) MSc (Institute of Cancer and Genomic Sciences, University of Birmingham, UK) & [Tasanee Braithwaite](mailto:tasaneebraithwaite@gmail.com) DM (The Medical Eye Unit, St Thomas’ Hospital NHS Foundation Trust, London, UK).
  * A.3 [Licence](https://github.com/sap218/ocimido/blob/master/LICENSE), which governs the permissions surrounding the ontology, https://creativecommons.org/licenses/by/3.0/.
  * A.4 Located in `ontology` directory: https://github.com/sap218/ocimido/blob/master/ontology/ocimido.owl.
  * A.5 GitHub repository: https://github.com/sap218/ocimido.
  * A.6 Methodological framework: used [Protégé](https://protege.stanford.edu/) software for ontology building and used the tf-idf statistical analysis to gain patient-preferred synonyms.

**B. Motivation**
  * B.1 Ontology is required as there currently lacks controlled vocabulary for ocular immune-mediated inflammatory diseases & the need for understanding the “patient-voice” [2].
  * B.2 OcIMIDo is derived from other biomedical ontologies, which are currently limited in the ocular domain, such as DOID, HPO, and ORDO.
  * B.3 Target audience are those in the ophthalmology research area, or even extended. In additional to those who wish to create an ontology in the method we have: the quick tf-idf method.

**C. Scope, requirements, development community**
  * C.1 The requirements of the ontology is to be an extensive vocabulary for ocular-related disorders and their associated diseases. The ontology includes therapeutic interventions, investigations, symptoms, complications, genotypic data, and more.
  * C.2 Development community: [Samantha C Pendleton](mailto:samanfapc@gmail.com) MSc (Institute of Cancer and Genomic Sciences, University of Birmingham, UK) & [Tasanee Braithwaite](mailto:tasaneebraithwaite@gmail.com) DM (The Medical Eye Unit, St Thomas’ Hospital NHS Foundation Trust, London, UK). In addition to Luke T Slater PhD, Alastair K Denniston PhD, and Georgios V Gkoutos PhD.
  * C.3 Issue tracking system used for future developements/requests/bugs: https://github.com/sap218/ocimido/issues 

**D. Knowledge acquisition**
  * D.1 Foundation built from medical document, Royal College of Ophthalmology Consensus document for Uveitis [3], and experts derived list of terms to be included – and then an extensive review of current biomedical ontologies
  * D.2 Medical document source https://www.rcophth.ac.uk/wp-content/uploads/2014/12/Uveitis-Data-Set-Oct-2018.pdf [3], biomedical ontologies were accessed via the ontology lookup service: https://www.ebi.ac.uk/ols/index [4].
  * D.3 Experts decided on the contents of OcIMIDo from the biomedical ontologies by searching through and extracting most relevant terms, for example, from UBERON we extracted eye-related anatomy concepts, e.g. `uvea` [UBERON:0001768]. However we avoided `iris nevus` [HP:0011525] from HPO.

**E. Ontology content**
  * E.1 Ontology is `OWL/XML xmlns= "http://www.w3.org/2002/07/owl#”` - ask creator if you wish for an alternative ontology format.
  * E.2 Specifically used [Protégé](https://protege.stanford.edu/) to develop the ontology with Git version control.
  * E.3 Ontology metrics, as of 14-06-2020 version 1.2.0 (see [`CHANGELOG.md`](https://github.com/sap218/ocimido/blob/master/CHANGELOG.md)) there are a total of 661 classes, 1661 relationships and axioms, 2851 annotations, including 1131 database cross-references, and 187 patient-preferred synonyms.
  * E.4 Cross-references ontologies inclue: DOID [5], HPO [6], ORDO [7], PATO [8], and UBERON [9].
  * E.5 IRI for OcIMIDo is: https://github.com/sap218/ocimido/blob/master/ontology/ocimido.owl#OCIMIDO_00000.
  * E.6 Identifier generation policy: Protégé’s preferences schema is that ontology’s entities are to have the prefix “OCIMIDO” followed by an underscore then a five digit number: currently they consists of 00001 up to ~000600. All classes have OCIMIDO prefix and own numeric identifier since OcIMIDo avoids importing axioms from other ontologies.
  * E.7 Entity metadata policy: currently `exact` synonyms in the ontology are additional clinical/medical terms. However `broad` synonyms are to be used for all Layman terms from the tf-idf method extracting from [Olivia's Vision](http://www.oliviasvision.org/) open forum.
  * E.8 BFO [10] and RO [11] were used as upper ontologies.
  * E.9 Currently, relationships in OcIMIDo are cross-referenced, such as `adjacent to` and `treatment of` from RO. We defined our own, `investigated by` (inverse `investigation for`) based on the investgations from the medical document [3].
  * E.10 Axiom patterns are inferred from the ontology structure itself (`subclass`) and by defined relationships, `is a` - currently only 100% true cases.

**F. Managing change**
  * F.1 Sustainability plan: OcIMIDo is freely available on GitHub and the creator, Samantha C Pendleton, will constantly update/change based on expert opinion or via Issue notifications. Anyone from the community can suggest new terms, but with sufficient evidence/justification. Potentially users can `fork` the ontology and make their own changes, which then allows the creator to merge these changes. If often particular individuals are making positive/appropriate changes they can be given direct access to co-maintain the ontology with the creator.
  * F.2 There is the change log ([CHANGELOG.md](https://github.com/sap218/ocimido/blob/master/CHANGELOG.md)) for entitles which need to be removed/split/redefined, additionally we can annotate these entities as obsolete. 
  * F.3 Ontology will follow “semantic versioning” guidelines and that it will maintain as version 1 unless a major foundation change is made then it will become version 2. Otherwise additional terms/fixes will increment to version 1.1 or 1.0.1 over time.

**G. Quality Assurance**
  * G.1 Successful testing on the Olivia’s Vision public forum for the `broad` synonyms plus annotations. Also used Protégé’s Pellet reasoner to ensure a coherent and consistent ontology. Experts in the opthamology background deemed/judged the ontology achieves the claims. 
  * G.2 Ontology is novel and first in ophthalmology. Additionally this method of building (tf-idf) is validated and proven a fast approach to synonym curation. OcIMIDo meets the stated requirements as an ontology suitable for the opthamology domain in additon to capturing the "patient voice.
  * G.3 An example of application is using the broad synonyms/Layman terms for sentiment analysis on the Olivia's Vision forum. Additionally future applications include GWAS/phenotype analysis on the UK Biobank cohort and other electronic health databases - OcIMIDo provides an implicit axiomatic structure for data annotated with SNOMED-CT, ICD-10, ICD-9, or Read codes as these are included as cross-references in the ontology, enhancing the potential value of OcIMIDo for searching and curating unstructured clinical data.
  * G.4 Ontology not yet available on OBO Foundry or other public repositories. However freely available on GitHub - ask creator if you wish for an alternative ontology format.
    5) G.5 Evidence of use* (MUST): An illustration of active projects and applications that use the ontology.
        ◦ Current evidence of OcIMIDo application is sentiment analysis on the Olivia’s Vision forum using the terms and synonyms for annotations, **note**: as soon as the manuscript is accepted the creator will include a reference/link here.

---

[1] Matentzoglu, N., Malone, J., Mungall, C., & Stevens, R. (2018). MIRO: guidelines for minimum information for the reporting of an ontology. Journal of Biomedical Semantics, 9(1), 6.

[2] Dean, S., Mathers, J. M., Calvert, M., Kyte, D. G., Conroy, D., Folkard, A., Southworth, S., Murray, P. I., & Denniston, A. K. (2017). “The patient is speaking”: discovering the patient voice in ophthalmology. The British Journal of Ophthalmology, 101(6), 700–708.

[3] Denniston, A. K., Lee, R. W., Pavesio, C., Stanford, M. R., Murray, P. I., Okada, A., Sen, H. N., & Dick, A. D. (2018). Uveitis Dataset Clinical Dataset. The Royal College of Ophthalmologists. https://www.rcophth.ac.uk/wp-content/uploads/2014/12/Uveitis-Data-Set-Oct-2018.pdf

[4] Côté, R. G., Jones, P., Apweiler, R., & Hermjakob, H. (2006). The Ontology Lookup Service, a lightweight cross-platform tool for controlled vocabulary queries. BMC Bioinformatics, 7, 97.

[5] Schriml, L. M., Arze, C., Nadendla, S., Chang, Y.-W. W., Mazaitis, M., Felix, V., Feng, G., & Kibbe, W. A. (2012). Disease Ontology: a backbone for disease semantic integration. Nucleic Acids Research, 40(Database issue), D940–D946.

[6] Robinson, P. N., Köhler, S., Bauer, S., Seelow, D., Horn, D., & Mundlos, S. (2008). The Human Phenotype Ontology: a tool for annotating and analyzing human hereditary disease. American Journal of Human Genetics, 83(5), 610–615.

[7] Vasant, D., Chanas, L., Malone, J., Hanauer, M., Olry, A., Jupp, S., Robinson, P. N., Parkinson, H., & Rath, A. (2014). Ordo: An ontology connecting rare disease, epidemiology and genetic data. Proceedings of ISMB, 30. 

[8] Gkoutos, G. V., Schofield, P. N., & Hoehndorf, R. (2018). The anatomy of phenotype ontologies: principles, properties and applications. Briefings in Bioinformatics, 19(5), 1008–1021.

[9] Mungall, C. J., Torniai, C., Gkoutos, G. V., Lewis, S. E., & Haendel, M. A. (2012). Uberon, an integrative multi-species anatomy ontology. Genome Biology, 13(1), R5.

[10] Smith, B., Kumar, A., & Bittner, T. (2005). Basic Formal Ontology for Bioinformatics. IFOMIS Reports.

[11] Smith, B., Ceusters, W., Klagges, B., Köhler, J., Kumar, A., Lomax, J., Mungall, C., Neuhaus, F., Rector, A. L., & Rosse, C. (2005). Relations in biomedical ontologies. Genome Biology, 6(5), R46.


CHANGELOG.md
## Change Log

* 07-01-2020 **version 1.0.0** - release v1.0.0

* 22-03-2020 **version 1.0.1** - all Layman terms (Olivia's Vision) are Broad Synonyms, the additional clinical terms (Royal College Clinical Dataset 2018) are Exact Synonyms to separate

* 23-03-2020 **version 1.0.2** - ICD-10 codes are now fixed (previously: H150, now: H15.0)

* 26-05-2020 **version 1.1.0** - newly added infectious and systemic diseases

* 29-05-2020 **version 1.1.1** - inflammatory disorders have ICD9 codes and readcodes [fully]

* 14-06-2020 **version 1.2.0** - all synonyms/xrefs are properly IRI linked to the Gene Ontology, e.g. http://www.geneontology.org/formats/oboInOWL#hasExactSynonym


text_mining/main.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Thu May  2 09:56:11 2019
@author: samantha
"""

import json
import re
from collections import Counter
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer
from nltk import ngrams

#-------------------- functions

def json_get(filename):
   with open(filename) as f_in:
       return(json.load(f_in)) # importing JSON
       
def get_all_id(posts):
    uids = []
    for post in posts:
        for comment in post:
            if "user_link" in comment:
                uids.append(comment["user_link"])
    return uids
       
def getting_uid(thread):
    user_link = thread[0].get('user_link')
    return user_link

def getting_user_posts_only(thread, user_link):
    if user_link is None:
        sentences = thread[0]['post_text']
        sentences = [re.sub('[^a-zA-Z0-9]+', ' ', _) for _ in sentences]
        sentences = ''.join(sentences)
    else:
        sentences = list(filter(lambda person: person['user_link'] == user_link, thread)) # getting persons' posts
        sentences = [" ".join(x["post_text"]) for x in sentences] # joining text
        sentences = [re.sub('[^a-zA-Z0-9]+', ' ', _) for _ in sentences] # removing special characters
        sentences = ' '.join(sentences) # joining 
        sentences = sentences.lower()
    return sentences

def nltk(string):
    stopWords = set(stopwords.words('english'))
    string = word_tokenize(string)
    wordsFiltered = [] # removing stop words
    for item in string:
        if item not in stopWords:
            wordsFiltered.append(item)
    stemming = [] 
    ps = PorterStemmer()
    for item in wordsFiltered:
        stemming.append(ps.stem(item))
    return stemming

def ngramming(string):
    bigramming = ngrams(string.split(), 2)
    trigramming = ngrams(string.split(), 3)
    bigrams = []
    for item in bigramming:
        bigrams.append(item)
    trigrams = []
    for item in trigramming:
        trigrams.append(item)
    return bigrams, trigrams

def type_of_uv(string):
    anterior = 0
    intermedi = 0
    posterior = 0
    panuv = 0
    if "anterior" in user_post:
        anterior += 1
    if "intermedi" in user_post:
        intermedi +=1
    if "posterior" in user_post:
        posterior += 1
    if "panuv" in user_post:
        panuv += 1
    type_of = (str(anterior) + str(intermedi) + str(posterior) + str(panuv))
    return type_of
    
#-------------------- code

if __name__ == "__main__":    
    posts = json_get('output.json') 
    
    uids = get_all_id(posts)
    uids = set(uids)
    
    df = pd.DataFrame()
    list_text = []
    list_uids = []
    list_type = []
        
    interesting_words = ["anterior", "intermedi", "posterior", "panuv"]
    for post in posts:
        #post = posts[p] # 13 ANON
        user_link = getting_uid(post)
        user_post = getting_user_posts_only(post, user_link) 
        user_post = nltk(user_post)
        user_post = ' '.join(user_post)
        
        list_text.append(user_post) # texts
        list_uids.append(user_link) # user_links
        list_type.append(type_of_uv(user_link)) # type of uveitis
    df = list(zip(list_text, list_uids, list_type))
    df = pd.DataFrame(df, columns = ['text' , 'uid', 'type']) 
    df.uid.replace([None], ("ANON"), inplace=True)
    #df['uid'].value_counts()
    list_strings = ' '.join(list_text)
    
    '''
    nonANONS = df
    nonANONS.uid.replace(["ANON"], ("0"), inplace=True)
    nonANONS = nonANONS[nonANONS['uid'] != "0"]
    newdf = nonANONS.groupby(['uid','type'])['text'].apply(', '.join).reset_index()
    newdf = newdf.groupby(['uid','type'])['text'].apply(', '.join).reset_index()
    '''
    
    
    #-------------------- type plot 
    types = []
    for item in df.type:
        types.append(item)
    types = dict(Counter(types))
    types = pd.DataFrame(list(types.items()), columns=['type', 'count'])
    types = types.sort_values('count', ascending=False)
    plt.figure(1)
    sns.barplot("type", "count", data=types).set_title("type of uveitis count")
    locs, labels = plt.xticks()
    plt.setp(labels, rotation=45)

    '''
    people_with = []
    people = (df.loc[df['type'] != '0000'])
    for item in people.uid:
        people_with.append(item)
    while "ANON" in people_with:
        people_with.remove("ANON")
    people_with = set(people_with)
    del people, item
    '''
    
    '''
    people_without = []
    people = (df.loc[df['type'] == '0000'])
    for item in people.uid:
        people_without.append(item)
    #while "ANON" in people_without:
    #    people_without.remove("ANON")
    people_without = set(people_without)
    del people, item
    '''
    
    #-------------------- grams
    bigrams, trigrams = ngramming(list_strings)
    popular_bigrams = dict(Counter(bigrams))
    popular_trigrams = dict(Counter(trigrams))
    del bigrams, trigrams    

    popular_bigrams = pd.DataFrame(list(popular_bigrams.items()), columns=['bigrams', 'count'])
    popular_bigrams = popular_bigrams.sort_values('count', ascending=False)
    popular_bigrams = popular_bigrams[popular_bigrams['count'] >= 20]
    popular_trigrams = pd.DataFrame(list(popular_trigrams.items()), columns=['trigrams', 'count'])
    popular_trigrams = popular_trigrams.sort_values('count', ascending=False)
    popular_trigrams = popular_trigrams[popular_trigrams['count'] >= 5]


    #-------------------- popular
    fig_size = plt.rcParams["figure.figsize"]
    fig_size[0] = 30
    fig_size[1] = 15
    plt.rcParams["figure.figsize"] = fig_size
    
    list_strings = list_strings.split()

    popular_words = dict(Counter(list_strings))
    
    popular = pd.DataFrame(list(popular_words.items()), columns=['words', 'count'])
    popular = popular.sort_values('count', ascending=False)
    popular = popular[popular['count'] >= 100]

    plt.figure(2)
    sns.barplot("words", "count", data=popular).set_title("popular word count")
    locs, labels = plt.xticks()
    plt.setp(labels, rotation=45)
    
    '''
    df.type.replace(("1000", "0100", "0010", "0001",
                     "1100", "1010", "1001",
                     "0101","0110",
                     "0011",
                     "1110",
                     "1111"), 
                    (interesting_words[0],interesting_words[1],interesting_words[2],interesting_words[3],
                     interesting_words[0]+interesting_words[1],interesting_words[0]+interesting_words[2],interesting_words[0]+interesting_words[3],
                     interesting_words[1]+interesting_words[2],interesting_words[1]+interesting_words[3],
                     interesting_words[2]+interesting_words[3],
                     interesting_words[0]+interesting_words[1]+interesting_words[2],
                     interesting_words[0]+interesting_words[1]+interesting_words[2]+interesting_words[3]), 
                    inplace=True)
    df['type'].value_counts()    
    '''    

text_mining/person_all_posts.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Apr 16 13:09:52 2019
@author: samantha

ALL posts of an individual user
"""

#-------------------- packages

import json
import re
from collections import Counter
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

#fig_size = plt.rcParams["figure.figsize"]
#fig_size[0] = 30
#fig_size[1] = 15
#plt.rcParams["figure.figsize"] = fig_size

from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer
from nltk import ngrams

#-------------------- functions
def js_r(filename):
   with open(filename) as f_in:
       return(json.load(f_in))
       
def get_all_id(posts):
    uids = []
    for post in posts:
        for comment in post:
            if "user_link" in comment:
                uids.append(comment["user_link"])
    return uids
def uids_post_count(uids):
    count = dict(Counter(uids))
    return count

def get_all_comments(posts, uid):
    comments = []
    for post in posts:
        for comment in post:
            if uid in comment.values():
                if "post_text" in comment:
                    comments.append(comment["post_text"])
    comments = [item for sublist in comments for item in sublist]
    comments = [re.sub('[^a-zA-Z0-9]+', ' ', _) for _ in comments] # removing special characters
    comments = ''.join(comments) # joining 
    comments = comments.lower()
    return comments

def nltk(string):
    stopWords = set(stopwords.words('english'))
    string = word_tokenize(string)
    wordsFiltered = [] # removing stop words
    for item in string:
        if item not in stopWords:
            wordsFiltered.append(item)
    stemming = [] 
    ps = PorterStemmer()
    for item in wordsFiltered:
        stemming.append(ps.stem(item))
    stemming = ' '.join(stemming)
    return stemming

def ngramming(string):
    bigramming = ngrams(string.split(), 2)
    trigramming = ngrams(string.split(), 3)
    bigrams = []
    for item in bigramming:
        bigrams.append(item)
    trigrams = []
    for item in trigramming:
        trigrams.append(item)
    return bigrams, trigrams

def type_of_uv(user_post):
    anterior = comments.count("anterior")
    intermedi = comments.count("intermedi")
    posterior = comments.count("posterior")
    panuv = comments.count("panuv")
    type_of = (str(anterior) + str(intermedi) + str(posterior) + str(panuv))
    return type_of

def getting_popular(comments):
    bigrams, trigrams = ngramming(comments)
    popular_bigrams = dict(Counter(bigrams))
    popular_trigrams = dict(Counter(trigrams))
    
    comments = comments.split()
    popular_words = dict(Counter(comments))
    
    return popular_words, popular_bigrams, popular_trigrams

#-------------------- code

if __name__ == "__main__": 
    posts = js_r('output.json')  
    
    uids = get_all_id(posts)
    uid_post_count = uids_post_count(uids) # getting user comments count
    
    uids = list(set(uids))
    uids = [x for x in uids if x is not None] # removing NULL uid
    
    list_uid = []
    list_comments = []
    list_types = []
    list_bigrams = []
    list_trigrams = []
    
    for uid in uids:
        #uid = uids[55] # using this UID | 55 & 34
        comments = get_all_comments(posts, uid) # users' all comments
    
        comments = nltk(comments) # NLTK
    
        #popular_words, popular_bigrams, popular_trigrams = getting_popular(comments)
        type_of = type_of_uv(comments)
    
        list_uid.append(uid)
        list_comments.append(comments)
        list_types.append(type_of)

        bigrams, trigrams = ngramming(comments)
        list_bigrams.append(bigrams)
        list_trigrams.append(trigrams)
    

    '''
    stemming = pd.DataFrame(list(stemming.items()), columns=['words', 'count'])
    stemming = stemming.sort_values('count')    
    plt.figure(1)
    sns.barplot("words", "count", data=stemming).set_title(user)
    locs, labels = plt.xticks()
    plt.setp(labels, rotation=90)
    plt.savefig("plots/UID:%s_top-words.png" % (user))    
    '''


text_mining/type_uveitis_all.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Thu May  2 10:35:20 2019
@author: samantha

TYPE of Uveitis of users
"""

from collections import Counter
import re
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer
import json

#-------------------- importing JSON
def js_r(filename):
   with open(filename) as f_in:
       return(json.load(f_in))
posts = js_r('output.json')

#-------------------- getting list of all users
users = []
for post in posts:
    for comment in post:
        if "user_link" in comment:
            users.append(comment["user_link"])

users = [x for x in users if x is not None]
uids = dict(Counter(users))
uids = list(uids)

#-------------------- getting list of comments for particular user
anterior = 0
immediate = 0
posterior = 0
panuveitis = 0

for item in uids:
    comments = []
    for post in posts:
        for comment in post:
            if item in comment.values():
                if "post_text" in comment:
                    comments.append(comment["post_text"])

    comments = [item for sublist in comments for item in sublist]
    comments = [re.sub('[^a-zA-Z0-9]+', ' ', _) for _ in comments] # removing special characters
    comments = ''.join(comments) # joining 
    comments = comments.lower()

    #-------------------- nltk
    stopWords = set(stopwords.words('english'))
    comments = word_tokenize(comments)
    wordsFiltered = [] # removing stop words
    for item in comments:
        if item not in stopWords:
            wordsFiltered.append(item)
    stemming = [] 
    ps = PorterStemmer()
    for item in wordsFiltered:
        stemming.append(ps.stem(item))
    string = stemming 
    string = ' '.join(map(str, string)) 
    
    #-------------------- type
    
    if "anteri" in string:
        anterior += 1
    if "intermedi" in string:
        immediate += 1
    if "posteri" in string:
        posterior += 1
    if "panuveit" in string:
        panuveitis += 1
    
#-------------------- counting
all_count = len(uids)
print("Total\tUSERS\t\t%s" % (all_count))
print("Total\tANTERIOR\t%s\t\t16.585365853658537" % (anterior))
print("Total\tINTERMEDIATE\t%s\t\t9.268292682926829" % (immediate))
print("Total\tPOSTERIOR\t%s\t\t10.24390243902439" % (posterior))
print("Total\tPANUVEITIS\t%s\t\t5.365853658536586" % (panuveitis))

count = anterior + immediate + posterior + panuveitis
print("Total\tUNKNOWN\t\t%s\t\t58.536585365853654" % (all_count-count))


text_mining/person_original_post.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Apr 16 13:09:52 2019
@author: samantha

ORIGINAL post of user
"""

from collections import Counter
import matplotlib.pyplot as plt
import pandas as pd
import re
import seaborn as sns

fig_size = plt.rcParams["figure.figsize"]
fig_size[0] = 30
fig_size[1] = 15
plt.rcParams["figure.figsize"] = fig_size

from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer
from nltk import ngrams

import json

def js_r(filename):
   with open(filename) as f_in:
       return(json.load(f_in)) # importing JSON
posts = js_r('output.json')

#-------------------- post check
print("Total count of posts\t1-%s" % (len(posts)))
query = input("What post do you wish to check?\t")
while int(query) > 416 or int(query) == 0:
    query = input("What post do you wish to check?\t")
query = int(query)-1    
post = posts[query]

#-------------------- users' post
try:
    user_link = [x["user_link"] for x in post]
    user_link = user_link[0]
    uid = [re.sub('[^a-zA-Z0-9]+', ' ', _) for _ in user_link]
    uid = uid[-10:]
    uid = ''.join(uid)
    uid = uid.replace(" ", "")
except TypeError:
        uid = "ANON"

#-------------------- getting user posts
person_posts = list(filter(lambda person: person['user_link'] == user_link, post)) # getting persons' posts
person_posts = [" ".join(x["post_text"]) for x in person_posts] # joining text
person_posts = [re.sub('[^a-zA-Z0-9]+', ' ', _) for _ in person_posts] # removing special characters
person_posts = ' '.join(person_posts) # joining 

#-------------------- stop words
stopWords = set(stopwords.words('english'))
person_posts = word_tokenize(person_posts)
wordsFiltered = [] # removing stop words
for item in person_posts:
    if item not in stopWords:
        wordsFiltered.append(item)
del stopWords

#-------------------- shortening & plotting
stemming = [] 
ps = PorterStemmer()
for item in wordsFiltered:
    stemming.append(ps.stem(item))
string = stemming
stemming = dict(Counter(stemming))
stemming = {k: v for k, v in stemming.items()} # if v > 1
stemming = pd.DataFrame(list(stemming.items()), columns=['words', 'count'])
stemming = stemming.sort_values('count')

plt.figure(1)
sns.barplot("words", "count", data=stemming).set_title(uid)
locs, labels = plt.xticks()
plt.setp(labels, rotation=90)
plt.savefig("plots/P:%s_UID:%s_top-words.png" % ((query+1), uid))


#-------------------- NGRAMS
string = ' '.join(map(str, string)) 
n = int(input("NGRAM\t(2 / 3):\t\t"))
sixgrams = ngrams(string.split(), n)
ngrams = []
for item in sixgrams:
    ngrams.append(item)
ngrams = dict(Counter(ngrams))
norg = ngrams
ngrams = {k: v for k, v in ngrams.items() if v > 1}

if len(ngrams) < 2:
    print("\nNo major NGRAMS\n")
    pass
    #for item in norg:
    #    print(item)
else:
    ngrams = pd.DataFrame(list(ngrams.items()), columns=['combinations', 'count'])
    ngrams = ngrams.sort_values('count')
    plt.figure(2)
    sns.barplot("combinations", "count", data=ngrams).set_title(uid)
    locs, labels = plt.xticks()
    plt.setp(labels, rotation=90)
    plt.savefig("plots/P:%s_UID:%s_ngrams_%s.png" % ((query+1), uid, n))
    

#-------------------- type

if "anteri" in string:
    print("type known: anterior")
if "intermedi" in string:
    print("type known: intermediate")
if "posteri" in string:
    print("type known: posterior")
if "panuveit" in string:
    print("type known: panuveitis")


text_mining/posts_all.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Mar 26 15:57:12 2019
@author: samantha

ALL results from every post/comment
"""

#------------------- importing modules
from collections import Counter
import matplotlib.pyplot as plt
import pandas as pd
import re
import numpy as np

import nltk
nltk.download('stopwords')
nltk.download('punkt')
nltk.download('wordnet')
from nltk.tokenize import sent_tokenize, word_tokenize, PunktSentenceTokenizer
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer, WordNetLemmatizer
from nltk.corpus import wordnet
from nltk.stem.wordnet import WordNetLemmatizer
from nltk import word_tokenize, pos_tag
from collections import defaultdict
from nltk import ngrams

###########################################################

file = "output.csv"
df = pd.read_csv(file)

x = 0
for item in range(16):
    df = df.drop((str(item) + "/post_date"), axis=1)
    df = df.drop((str(item) + "/topic_id"), axis=1)
    df = df.drop((str(item) + "/user_link"), axis=1)
    df = df.drop((str(item) + "/user_name"), axis=1)
    df = df.drop((str(item) + "/topic_title"), axis=1) ## BLAH
    x = x+1

columns = []
for row in df.iterrows():
    index, data = row
    columns.append(data.tolist())

flat_list = [str(item).lower() for sublist in columns for item in sublist]
flat_list = [x for x in flat_list if str(x) != 'nan']


column_split = []
for item in flat_list:
    column_split.append(item.split())

flat_list = [str(item).lower() for sublist in column_split for item in sublist] # POSTS ONLY 

###########################################################

df = pd.read_csv(file)
column = list(df['0/topic_title'])
column_split = []
for item in column:
    column_split.append(item.split())
flat_list2 = [item.lower() for sublist in column_split for item in sublist]


flat_list3 = flat_list + flat_list2 # combining posts and topic titles

###########################################################
###########################################################
###########################################################


flat_list3 = [re.sub('[^a-zA-Z0-9]+', '', _) for _ in flat_list3] # removing special chars
flat_list4 = [x for x in flat_list3 if x] # removing empty items

flat_list5 = flat_list4
string = ' '.join(map(str, flat_list5))

###########################################################
###########################################################

# removing stop words - a/the/this
stopWords = set(stopwords.words('english'))
words = word_tokenize(string)
wordsFiltered = [] # removing stop words
for w in words:
    if w not in stopWords:
        wordsFiltered.append(w)

###########################################################

stemming = [] # shorten words
ps = PorterStemmer()
for w in wordsFiltered:
    stemming.append(ps.stem(w))

stemmed = [] # shorten words
ps = PorterStemmer()
for w in stemming:
    stemmed.append(ps.stem(w))

###########################################################
###########################################################
###########################################################

all_words = dict(Counter(stemmed))
words = {k: v for k, v in all_words.items() if v > 60} 

bad_ones = ['1', '2', '3', '4', '5', '6','8','10',
        'abl', 'advic','ago','alreadi','also','although','alway','anoth','answer','anyth','around','ask','actual','anyway','ayliff',
        'becca','becom','bristol','believ',
        'call','cant','caron','child','children','christma','come','could','came','consid','chanc','contact',
        'daughter','day','didnt','discuss','doesnt','dont','dr','difficult','decid',
        'enough','ever','everi','everyon','everyth','etc','earli','especi','expect',
        'follow','fine','free','friend','forward',
        'get','give','given','go','goe','got','gone','gener',
        'happen','happi','help','hi','hope','howev','hello','havent','he',
        'id','idea','ill','im','inform','ive','isnt','info',
        'jack','jia','j',
        'keep','know','kid',
        'last','let','like','link','lot','love','luck','least','later','low','local',
        'made','make','mani','maxidex','may','mean','mention','methotrex','might','month','much','moment','moorfield','mind',
        'need','never','new','news','next','nicki','noth','notic','nice','nur','nh',
        'ok','old','olivia','other','owen','offer',
        'peopl','point','posit','possibl','problem','prof','put','place','parent','person','probabl','privat','pretti',
        'question','quit','quiet',
        'read','realli','riley','reason','recent','rather','regard',
        'said','say','school','seem','seen','sinc','someth','son','soon','sorri','sound','stay','still','stop','sure','sarah','someon','sometim','she','saw','sort','sam','stori','suggest','speak','sherri',
        'talk','tell','thank','that','thing','think','though','thought','told','today','tomorrow','tri','turn',
        'understand','us','use','updat','usual','u',
        'visit',
        'wait','want','way','week','well','went','whether','wish','wor','wont','what','wasnt','without','wonder','work','worri','would','worth','websit',
        'x','xx',
        'ye','yet','your','youv','yesterday','year',
        ]
for item in bad_ones:
    del words[item]

fig_size = plt.rcParams["figure.figsize"]
fig_size[0] = 35
fig_size[1] = 17
plt.rcParams["figure.figsize"] = fig_size

import seaborn as sns
data = pd.DataFrame(list(words.items()), columns=['word', 'count'])
data = data.sort_values('count')
plt.figure(1)
sns.barplot("word", "count", data=data)
locs, labels = plt.xticks()
plt.setp(labels, rotation=90)
plt.savefig("plots/ALL.png")

###########################################################
###########################################################

string = ' '.join(map(str, stemmed)) # NGRAMS
bad_ones = ['hi','edit','dont','know','fort','finger','back','nicki','sure','luck','next','maxin','tania',
            'news','post','like','hope','wish','thank','sorri','becca','thank','anyon','realli','daughter',
            'glad','everi','though','might','keep','olivia','plea','someth','would','goe','hate','william',
            'mistrust','penyebab','dedic','gejala','inform','wont','darah','read','tinggi','asam','websit']
for item in bad_ones:
    string = string.replace(item, '')
n = int(input("NGRAM\t(2 / 3):\t\t"))
#n = 3
sixgrams = ngrams(string.split(), n)
ngrams = []
for grams in sixgrams:
  ngrams.append(grams)
ngrams = dict(Counter(ngrams))
if n == 2:
    ngrams = {k: v for k, v in ngrams.items() if v > 20}
elif n == 3:
    ngrams = {k: v for k, v in ngrams.items() if v > 5}

ngrams = pd.DataFrame(list(ngrams.items()), columns=['combinations', 'count'])
ngrams = ngrams.sort_values('count')
#print(ngrams)
plt.figure(2)
sns.barplot("combinations", "count", data=ngrams)
locs, labels = plt.xticks()
plt.setp(labels, rotation=90)
plt.savefig("plots/ALL_ngrams_%s.png" % (n))